src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamValueTaskSource.cs

```diff
@@ -0,0 +1,252 @@
+                            new MemoryAwaitableProvider(strategy, numBufferedBytes, memory);
+            }
+
+            protected FileStreamValueTaskSource(
+                AsyncWindowsFileStreamStrategy strategy,
+                PreAllocatedOverlapped? preallocatedOverlapped,
+                int numBufferedBytes,
+                byte[]? bytes)
+            {
+                _strategy = strategy;
+                _numBufferedBytes = numBufferedBytes;
+
+                _result = FileStreamHelpers.NoResult;
+
+                _source = default;
+                // Using RunContinuationsAsynchronously for compat reasons (old API used Task.Factory.StartNew for continuations)
+                _source.RunContinuationsAsynchronously = true;
+
+                _overlapped = bytes != null &&
+                              _strategy.CompareExchangeCurrentOverlappedOwner(this, null) == null ?
+                              _strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(preallocatedOverlapped!) : // allocated when buffer was created, and buffer is non-null
+                              strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(s_ioCallback, this, bytes);
+
+                Debug.Assert(_overlapped != null, "AllocateNativeOverlapped returned null");
+            }
+
+            internal NativeOverlapped* Overlapped => _overlapped;
+            internal int NumBufferedBytes => _numBufferedBytes;
+            public ValueTaskSourceStatus GetStatus(short token) => _source.GetStatus(token);
+            public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) => _source.OnCompleted(continuation, state, token, flags);
+            void IValueTaskSource.GetResult(short token) => _source.GetResult(token);
+            int IValueTaskSource<int>.GetResult(short token) => _source.GetResult(token);
+            internal short Version => _source.Version;
+
+            internal void RegisterForCancellation(CancellationToken cancellationToken)
+            {
+#if DEBUG
+                Debug.Assert(cancellationToken.CanBeCanceled);
+                Debug.Assert(!_cancellationHasBeenRegistered, "Cannot register for cancellation twice");
+                _cancellationHasBeenRegistered = true;
+#endif
+
+                // Quick check to make sure the IO hasn't completed
+                if (_overlapped != null)
+                {
+                    // Register the cancellation only if the IO hasn't completed
+                    long packedResult = Interlocked.CompareExchange(ref _result, FileStreamHelpers.RegisteringCancellation, FileStreamHelpers.NoResult);
+                    if (packedResult == FileStreamHelpers.NoResult)
+                    {
+                        _cancellationRegistration = cancellationToken.UnsafeRegister(static (s, token) => ((FileStreamValueTaskSource)s!).Cancel(token), this);
```

Comment:
Could we pass the callback on a similar fashion to FileStreamCompletionSource 
https://github.com/dotnet/runtime/blob/57bbdee53593fb89913c5f2e19fdcdc10eac3980/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamCompletionSource.Win32.cs#L92

Or why did you opted for this way instead?

Reply:
It makese more sense to pass `this` as the second argument of `UnsafeRegister`, which ensures it's a `ValueTaskSource` instance, and saves us from the check of the `object? state` parameter like in the old code. The old code [also allocates](https://github.com/dotnet/runtime/blob/a913740aac5353212a9b95283ae38b9964f18aa5/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamCompletionSource.Win32.cs#L86) an `Action<object?>` instance to save a reference to the `Cancel` method.

We have a [similar example](https://source.dot.net/#System.Threading.Channels/System/Threading/Channels/AsyncOperation.cs,92) in System.Threading.Channels.

Reply:
> It made more sense to pass this as the second argument of UnsafeRegister

FileStreamCompletionSource was also passing `this` as 2nd argument.

> The old code also allocates an Action<object?> instance to save a reference to the Cancel method.

Yes, seems that it was caching the method group to avoid allocating everytime it was passed to `UnsafeRegister`.
However I think you can avoid the allocation without caching yourself by doing this:
```cs
_cancellationRegistration = cancellationToken.UnsafeRegister((s, token) => Cancel(s, token), this);
```

Just a suggestion, not something that you should block on.